File: kotlin/advent/src/main/kotlin/org/gristle/pdxpuzzles/advent/y2016/Y16D01.kt
package org.gristle.pdxpuzzles.advent.y2016

import org.gristle.pdxpuzzles.advent.utilities.Day
import org.gristle.pdxpuzzles.utilities.objects.Coord
import org.gristle.pdxpuzzles.utilities.enums.Nsew
import org.gristle.pdxpuzzles.utilities.parsing.getInts

class Y16D01(input: String) : Day {

    private val moves: Sequence<Coord>

    init {
        var dir = Nsew.NORTH

        moves = input
            .splitToSequence(", ")
            .flatMap { instruction ->
                dir = if (instruction[0] == 'L') dir.left() else dir.right()
                List(instruction.getInts().first()) { dir }
            }.runningFold(Coord.ORIGIN, Coord::move)
    }

    override fun part1() = moves
        .last()
        .manhattanDistance()

    override fun part2(): Int {
        val visited = mutableSetOf<Coord>()
        return moves
            .first { !visited.add(it) }
            .manhattanDistance()
    }
}

fun main() = Day.runDay(Y16D01::class)

//    Class creation: 6ms
//    Part 1: 226 (3ms)
//    Part 2: 79 (4ms)
//    Total time: 14ms

File: rust/advent/src/bin/2016/y16d01.rs
use advent::utilities::get_input::get_input;
use rustc_hash::FxHashSet;
use utilities::{enums::cardinals::Cardinal, structs::{coord::Coord2, stopwatch::{ReportDuration, Stopwatch}}};

type Input = Vec<Pos>;
type Output = usize;
type Pos = Coord2;

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let input = get_input(16, 1).unwrap();
    let input = parse_input(&input);
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input), stopwatch.lap().report());
    println!("2. {} ({})", part2(&input), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

fn parse_input(input: &str) -> Input {
    let mut dir = Cardinal::North;
    input.split(", ")
        .flat_map(|instruction| {
            dir = if instruction.starts_with('L') { dir.left() } else { dir.right() };
            let dist = instruction[1..].parse().unwrap();
            std::iter::repeat(dir).take(dist)
        })
        .scan(Pos::origin(), |pos, dir| {
            *pos = pos.move_direction(dir, 1).unwrap();
            Some(*pos)
        })
        .collect()
}

fn part1(moves: &Input) -> Output {
    moves.last().unwrap().manhattan_distance(Pos::origin())
}

fn part2(moves: &Input) -> Output {
    let mut visited = FxHashSet::default();
    moves.iter()
        .find(|&&pos| !visited.insert(pos))
        .map(|pos| pos.manhattan_distance(Pos::origin()))
        .unwrap()
}

#[test]
fn default() {
    let input = get_input(16, 1).unwrap();
    let input = parse_input(&input);
    assert_eq!(226, part1(&input));
    assert_eq!(79, part2(&input));
}

// Input parsed (35μs)
// 1. 226 (6μs)
// 2. 79 (24μs)
// Total: 68μs
